src/native/aws_signing.c

```diff
@@ -0,0 +1,325 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "crt.h"
+#include "http_request_utils.h"
+#include "java_class_ids.h"
+
+#include <jni.h>
+#include <string.h>
+
+#include <aws/auth/credentials.h>
+#include <aws/auth/signable.h>
+#include <aws/auth/signing.h>
+#include <aws/common/string.h>
+#include <aws/http/request_response.h>
+
+/* on 32-bit platforms, casting pointers to longs throws a warning we don't need */
+#if UINTPTR_MAX == 0xffffffff
+#    if defined(_MSC_VER)
+#        pragma warning(push)
+#        pragma warning(disable : 4305) /* 'type cast': truncation from 'jlong' to 'jni_tls_ctx_options *' */
+#    else
+#        pragma GCC diagnostic push
+#        pragma GCC diagnostic ignored "-Wpointer-to-int-cast"
+#        pragma GCC diagnostic ignored "-Wint-to-pointer-cast"
+#    endif
+#endif
+
+struct s_aws_sign_request_callback_data {
+    JavaVM *jvm;
+    jobject java_future;
+    jobject java_original_request;
+    jobject java_sign_param_predicate;
+    struct aws_http_message *native_request;
+    struct aws_signable *original_message_signable;
+    struct aws_string *region;
+    struct aws_string *service;
+};
+
+static void s_cleanup_callback_data(struct s_aws_sign_request_callback_data *callback_data) {
+
+    JNIEnv *env = aws_jni_get_thread_env(callback_data->jvm);
+
+    (*env)->DeleteGlobalRef(env, callback_data->java_future);
+    (*env)->DeleteGlobalRef(env, callback_data->java_original_request);
+
+    if (callback_data->java_sign_param_predicate) {
+        (*env)->DeleteGlobalRef(env, callback_data->java_sign_param_predicate);
+    }
+
+    if (callback_data->native_request) {
+        aws_http_message_release(callback_data->native_request);
+    }
+
+    if (callback_data->original_message_signable) {
+        aws_signable_destroy(callback_data->original_message_signable);
+    }
+
+    aws_string_destroy(callback_data->region);
+    aws_string_destroy(callback_data->service);
+
+    aws_mem_release(aws_jni_get_allocator(), callback_data);
+}
+
+static jobject s_create_signed_java_http_request(
+    JNIEnv *env,
+    struct aws_http_message *native_request,
+    jobject java_original_request) {
+    jstring jni_uri = NULL;
+    jobjectArray jni_headers = NULL;
+    jobject http_request = NULL;
+
+    jstring jni_method = (*env)->GetObjectField(env, java_original_request, http_request_properties.method_field_id);
```

Comment:
since jni_method and jni_uri are seated on line 104, should we be calling DeleteLocalRef on those as well?

Reply:
Yes